src/strawberry_sqlalchemy_mapper/field.py

```diff
@@ -0,0 +1,731 @@
+from __future__ import annotations
+
+import asyncio
+import contextlib
+import contextvars
+import dataclasses
+import inspect
+from collections import defaultdict
+from typing import (
+    Any,
+    Callable,
+    DefaultDict,
+    Dict,
+    Generator,
+    Iterable,
+    Iterator,
+    List,
+    Literal,
+    Mapping,
+    Optional,
+    Sequence,
+    Tuple,
+    Type,
+    TypeVar,
+    Union,
+    cast,
+    overload,
+)
+from typing_extensions import Annotated, TypeAlias
+
+from sqlakeyset.types import Keyset
+from sqlalchemy.ext.asyncio import AsyncSession
+from sqlalchemy.orm import Query, Session
+from strawberry import relay
+from strawberry.annotation import StrawberryAnnotation
+from strawberry.arguments import StrawberryArgument, argument
+from strawberry.extensions.field_extension import (
+    FieldExtension,
+)
+from strawberry.field import (
+    _RESOLVER_TYPE,
+    StrawberryField,
+)
+from strawberry.permission import BasePermission
+from strawberry.relay.exceptions import RelayWrongAnnotationError
+from strawberry.relay.types import NodeIterableType
+from strawberry.type import (
+    StrawberryList,
+    StrawberryOptional,
+    get_object_definition,
+)
+from strawberry.types import Info
+from strawberry.types.fields.resolver import StrawberryResolver
+from strawberry.utils.aio import asyncgen_to_list
+
+_T = TypeVar("_T")
+_SessionMaker: TypeAlias = Callable[[], Union[Session, AsyncSession]]
+
+assert argument  # type: ignore[truthy-function]
+
+
+connection_session: contextvars.ContextVar[
+    Union[Session, AsyncSession, None]
+] = contextvars.ContextVar(
+    "connection-session",
+    default=None,
+)
+
+
+@contextlib.contextmanager
+def set_connection_session(
+    s: Union[Session, AsyncSession, None]
+) -> Generator[None, None, None]:
+    token = connection_session.set(s)
+    try:
+        yield
+    finally:
+        connection_session.reset(token)
+
+
+class StrawberrySQLAlchemyField(StrawberryField):
+    """
+    Base field for SQLAlchemy types.
+    """
+
+    def __init__(
+        self,
+        sessionmaker: _SessionMaker | None = None,
+        keyset: Keyset | None = None,
```

Comment:
This is at field definition time, correct? Keyset I believe should come not at definition time but at query time. The Keyset gets passed to the caller as the pagination token (well technically the Marker gets passed, which includes the Keyset), which gets passed back when fetching the next page.

Reply:
@mattalbr not sure if I understood whast you mean, and maybe the name of the variable here is not the best one =P

This variable defines the fields that will be used for the pagination, i.e., that will be used to `order_by` the query when you are relying on the default resolver.

You can see an example in [this test](https://github.com/strawberry-graphql/strawberry-sqlalchemy/pull/65/files#diff-e466e4061d3ec646290e8d41bfb1ccfb9f3d7f36113d7c8c9196d2e1ae9de05aR509)

Reply:
@mattalbr oh I see. In the way I implemented it, the ordering is defined in the field

The old link is not working anymore, here is a new one: https://github.com/strawberry-graphql/strawberry-sqlalchemy/pull/65/files#diff-e466e4061d3ec646290e8d41bfb1ccfb9f3d7f36113d7c8c9196d2e1ae9de05aR509

Reply:
Oh, the link is not working because the file itself is collapsed.

Here is the screenshot of the implementation

![Screenshot 2023-11-30 at 18 12 31](https://github.com/strawberry-graphql/strawberry-sqlalchemy/assets/134025/a8066a16-bc6e-458d-af13-573f8f6ebb21)

Reply:
Having said that 

> Sorry that link isn't working for me, can you point me to the test by name? I just want to make sure that which fields will be used to order the query is defined per request. E.g. imagine we had a resolver:
> 
> ```python
> def fruits(order: list[string], info: Info):
>     return info.context.session.query(models.Fruit).order(*order).all()
> ```
> 
> The ordering, and thus the keyset, is not fixed for the resolver -- it's determined by the client.

Having said that, your example here would actually work. keyset is only used for the automatic resolver, not for manual ones. So if you define a resolver like you did it will work